### PR TITLE
M2/NV18 Support

### PR DESCRIPTION
**Pull request by vyzo**
This adds support for v10/nv18 actor states, in order to facilitate the transition of M2/NV18 branches into master.

TODO:
- [ ] NV18 migration; https://github.com/filecoin-project/go-state-types/issues/120
- [x] Copy code from v9 actors to v10 as requested by his yushiness.

### CODE REVIEW

**Comment by arajasek** [comment]
I haven't written this up yet, but I decided to avoid type aliasing for the v9 state types. I think I like them to be standalone types that just say "this is what state X looked like in Filecoin actors v9". 

specs-actors had aliased types, and in practice, this actually proved to be somewhat painful when it came to integrations. See, for example, the raft of PRs opened [this year](https://github.com/filecoin-project/specs-actors/pulls?q=is%3Apr+is%3Aclosed) that required v3 to be updated in v4 to be updated in v5, etc.

I also think it's led to cleaner code in Lotus with the v9 integration -- there's some type converting glue in the generated accessors, but everything else looks cleaner. We've updated to using the new v9 types where appropriate, and have been forced to introduce abstractions where necessary (instead of trying to shim things in the various use cases).

I'm not sold to this decision, so I'm happy to hear others thoughts here / more from you @vyzo. @Stebalien @ZenGround0 @geoff-vball @anorth seem like good people to poll as folks who worked on actors / have integrated things in Lotus.

> As v9 is still technically in development) 

Nah, it's done, any changes from this point better be few and small in scope (none expected).

**Comment by vyzo** [comment]
My main rationale is that we are not touching these actors at all, and blatant copy paste just feels wrong.

I am open to copying, but it gives me an allergic reaction.

**Comment by geoff-vball** [comment]
@arajasek I agree that copying the old actors and making the appropriate changes made integration relatively painless.

**Comment by vyzo** [comment]
Sure, but there are no changes whatsoever, except for the init actor which has been copied.

**Comment by geoff-vball** [comment]
We'll need to make more changes for nv18 depending on what FIPs are accepted. Having the copied versions already there for us to start working on will help :)

**Comment by vyzo** [comment]
We can always copy on demand.

**Comment by raulk** [comment]
Ok, we see things differently. Disagree and commit it is ;-)

**Comment by vyzo** [comment]
Ok, so what do? I am fine with removing the aliasing and copying, as much as I don't like it. But then we'll have to remember to update if we make changes in v9 actors.

**Comment by vyzo** [comment]
cherry picked the recent changes (from Lisbon rush) from experimental m2 branch.

**Comment by vyzo** [comment]
rebased on master

**Comment by vyzo** [comment]
Note for @arajasek and all the proponents of copying: aliasing allows me to simply rebase on master instead of painstakingly hunting down changes; how can you argue with that?

**Comment by vyzo** [comment]
Ok, will do once we make sure this is working fine and it is ready for merge.

**Comment by vyzo** [comment]
rebased on master.

File: builtin/v10/eam/eam_types.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+}
+
+type Create2Params struct {
+	Initcode []byte
+	Salt     [32]byte
+}
+
+type CreateAccount struct {
+	// Pubkey is the secp256k1 public key
+	Pubkey [65]byte
```
Comment (arajasek) [inline builtin/v10/eam/eam_types.go:17]: Can we define a SECP_PUBKEY_LENGTH or something somewhere?

File: builtin/v10/evm/evm_actor_state.go [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package evm
+
+import (
+	"github.com/ipfs/go-cid"
+)
+
+type State struct {
+	Bytecode      cid.Cid
+	ContractState cid.Cid
```
Comment (arajasek) [inline builtin/v10/evm/evm_actor_state.go:9]: Nonce?

**Review by arajasek — COMMENTED** [review]
Generally looks good, needs the un-aliasing of v9-v10.

File: builtin/v10/evm/evm_actor_state.go (hunk 2 above)
Comment (vyzo) [reply builtin/v10/evm/evm_actor_state.go:9]: it doesn't have such a field in its state!

Comment (vyzo) [reply builtin/v10/evm/evm_actor_state.go:9]: oh wait, it does. Recent addition, will fix.

File: builtin/v10/eam/eam_types.go (hunk 1 above)
Comment (vyzo) [reply builtin/v10/eam/eam_types.go:17]: surely.

**Review by Stebalien — DISMISSED** [review]

File: builtin/v10/evm/evm_actor_state.go (hunk 2 above)
Comment (Stebalien) [reply builtin/v10/evm/evm_actor_state.go:9]: Fixed in https://github.com/filecoin-project/go-state-types/pull/116.

**Comment by vyzo** [comment]
Your copy is ready @arajasek.

File: network/version.go [hunk 3]
```diff
@@ -25,6 +25,7 @@ const (
 	Version15                 // ???
 	Version16                 // ???
 	Version17                 // ???
+	Version18                 // FVM M2.1
```
Comment (arajasek) [inline network/version.go:28]: I'd suggest adding a couple more versions here just to have them ready for buildernet.

Also can you add the names for Versions 15-17 please?

15 is OhSnap, 16 is Skyr, 17 is Shark

File: builtin/v8/account/cbor_gen.go [hunk 4]
```diff
@@ -5,12 +5,16 @@ package account
...
 	"io"
+	"sort"
 
+	cid "github.com/ipfs/go-cid"
 	cbg "github.com/whyrusleeping/cbor-gen"
 	xerrors "golang.org/x/xerrors"
 )
 
 var _ = xerrors.Errorf
+var _ = cid.Undef
```
Comment (arajasek) [inline builtin/v8/account/cbor_gen.go:16]: Do you understand why these files are changing in this weird way? Guessing it has to do with the `github.com/whyrusleeping/cbor-gen` bump, but it'd be nice to be sure.

**Review by arajasek — APPROVED** [review]

File: network/version.go (hunk 3 above)
Comment (vyzo) [reply network/version.go:28]: sure

File: builtin/v8/account/cbor_gen.go (hunk 4 above)
Comment (vyzo) [reply builtin/v8/account/cbor_gen.go:16]: yrah, new version in town.
There shouldnt be any _functional_ changes.

File: network/version.go (hunk 3 above)
Comment (vyzo) [reply network/version.go]: done.

**Pull request merged**